### DISCUSSION TITLE
Setting image upload path dynamically

### DISCUSSION THREAD

**Discussion by developkariyer** [pimcore/pimcore#17126] [Admin-Ui-Classic-Bundle]
Hi,
Trying to migrate to PIMCore from google sheets for product and related asset management, working on a proof-of-concept, before decision on going enterprise.

I have an image component in my Product class. I want each image to be uploaded to a specific folder for that instance of Product (a path using product->productCode). Also, I need to see those uploaded images in Assets and also in my current cloud storage solution, which Pimcore uploaded files folder will be synced.

Currently, I failed to set a custom upload path for each Product. I added event listener for preAdd and preUpdate, I tried setUploadPath, setCustomUploadPath, setPath on product->picture (name of my image component), all failed with no such method.

So, question is, is it possible to set an upload path for each object of my Product class?

**Comment by cancan101** [comment]
Did you take a look at: https://pimcore.com/docs/platform/Pimcore/Extending_Pimcore/Event_API_and_Event_Manager/#asset-upload-path

**Comment by arulselvan008** [comment]
Recently, we achieved dynamic asset folder creation and asset renaming based on our custom logic using the DataObjectEvent in Pimcore.

We structured the asset hierarchy in our application around brand, category, and subcategory, creating folders accordingly. Images are named as follows:

Thumbnail Image Upload: When uploading a thumbnail, the image is named {ProductSKU}_T.ImageExtension (e.g., GSM_T.jpg).
Gallery Image Upload: Gallery images are named {ProductSKU}_H1.ImageExtension (e.g., GSM_H1.jpg), with _H2, _H3, etc., for subsequent images.
After updating the data object, we trigger the pimcore.dataobject.postUpdate method. This method renames and moves the asset from the default upload folder to our custom asset hierarchy, based on the data object details.

Please configure the /config/services.yaml file and create the AssetRenameListener.php file as shown in the sample below.

services:
    App\EventListener\AssetRenameListener:
        autowire: true
        autoconfigure: true
        tags:
            - { name: kernel.event_subscriber }

namespace App\EventListener;

use Pimcore\Event\Model\DataObjectEvent;
use Pimcore\Model\Asset;
use Pimcore\Model\DataObject\AllProducts;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;

class AssetRenameListener implements EventSubscriberInterface
{
    public static function getSubscribedEvents()
    {
        return [
            'pimcore.dataobject.postUpdate' => 'onDataObjectPostSave',
        ];
    }

    public function onDataObjectPostSave(DataObjectEvent $event)
    {
        $object = $event->getObject();
    
       // Skip if the data object is newly created (empty object) or auto-saved.
    
      // Your custom code implementation for asset renaming and handling.
      // Retrieve the current and previous published versions of the object to compare if there are any updates to the images.
      // Continue with your custom logic implementation here.
    }
}

**Comment by developkariyer** [comment]
Thanks for pointing the detail in the documentation, and also for sample code. 
I am following onPostXXX of object, which I already have some listeners.
Thanks for all.
